index.js

```diff
@@ -261,55 +298,84 @@ class ClientServiceBase {
 
     const builder = new ContractExecutionRequestBuilder(
         new this.protobuf.ContractExecutionRequest(),
-        this.signer).withContractId(contractId).
-        withContractArgument(argumentJson).
-        withFunctionArgument(functionArgumentJson).
-        withCertHolderId(this.certHolderId).
-        withCertVersion(this.certVersion);
+        this.signer,
+    ).withContractId(contractId)
+        .withContractArgument(argumentJson)
+        .withFunctionArgument(functionArgumentJson)
+        .withCertHolderId(this.certHolderId)
+        .withCertVersion(this.certVersion);
 
     const request = await builder.build();
+    const promise = new Promise((resolve, reject) => {
+      this.ledgerClient.executeContract(
+          request,
+          this.metadata,
+          (err, response) => {
+            if (err) {
+              reject(err);
+            } else {
+              const jsonResponse = response.toObject();
+              jsonResponse.result = JSON.parse(jsonResponse.result);
+              resolve(jsonResponse);
+            }
+          },
+      );
+    });
 
-    return this.sendRequest('executeContract', () =>
-      new Promise((resolve, reject) => {
-        this.ledgerClient.executeContract(request, this.metadata,
-            (err, response) => {
-              if (err) {
-                reject(err);
-              } else {
-                resolve(response);
-              }
-            });
-      }));
+    return this.sendRequest(() => promise);
   }
 
   /**
-   * @param {string} funcName
    * @param {Promise} func
    * @return {Promise}
+   * @throws {ClientError}
    */
-  async sendRequest(funcName, func) {
+  async sendRequest(func) {
```

Comment:
Sorry I'm not  sure about `js`, but is this meaningful to do `async` for this and doing `return await func()` in internal ? Just don't know the benefit.

Reply:
The method name is changed so it is outdated, but i'm still wondering. 
Is this just a best practice in js ? (Sorry for a dumb question...)

Reply:
Sorry, it is really hard to answer.
I will try to explain.

I revised the `sendRequest(func)` to `_executePromise(promise)` for better readability (I hope).
I can't remember why I used an anonymous function.
I think we don't really need it.

Back to the question.
Why do we need to do this:

```
async _executePromise(promise) {
    try {
      return await promise;
    } catch (e) {
      const status = this._parseStatusFromError(e);
      if (status) {
        throw new ClientError(status.code, status.message);
      } else {
        throw new ClientError(
            StatusCode.UNKNOWN_TRANSACTION_STATUS,
            e.message,
        );
      }
    }
}
```
It is all about to make asynchorize functions possible to work in order.

The underneath functions provided by the protobuf are asynchronized.
For example,
```
protobuf.registerContract(registerContractRequest, callback) {
   ...
}
```
and
```
protobuf.executeContract(executeContractRequest, callback) {
   ...
};
```

If we don't make them be able to be `synchronized`,
the way to make sure they are executed in order will be like

```
protobuf.registerContract(registerContractRequest, (error, response) => {
    if (error == null) {
        protobuf.executeContract(executeContractRequest, (error2, response2) => {
            if (error == null) {
                // handle response2
            }
        });
    }
});
```
It is very annoying as the code flow gets longer !
(Even more annoying in the era that didn't have anoynomous functions `() => {}`)

Therefore, JavaScript Promise comes to the play to improve readability.

Let's see Promise first.
In our example, first of all, we use Promise to `wrap` protobuf's functions


```
registerContractPromise = new Promise((resolve, reject) => {
    protobuf.registerContract(registerContractRequest, (error, response) => {
        if (error) {
            reject(error);
        } else {
            resolve(response);
        }
    }
})
```

```
executeContractPromise = new Promise((resolve, reject) => {
    protobuf.executeContract(executeContractRequest, (error, response) => {
        if (error) {
            reject(error);
        } else {
            resolve(response);
        }
    }
})
```

Then, we can use Promise in this way.

```
registerContractPromise
  .then(executeContractPromise)
  .then(executeContractResponse => {
    // handle executeContractResponse
  })
  .catch(error => {
    ...
  });
```

Hmm, ok ... a bit better. But, not that ideal.

The time came to ES5 (ES2016) era.
Language sugar async and await occurs.

The code using await looks like this

```
try {
    await registerContractPromise;
    const response = await executeContractPromise;
} catch (error => {
    console.log(error);
});

```
await will `execute` a Promise, wait for the result and proceed the flow.

It looks much better, doesn't it?

However, we have to understand, await is a language sugar after all.
Behind the sugar, Promise is still Promise.

That is why await always has to be inside async

```

async foo() {
    await promise1;
    await promise2;
    await promise3;
    return null;
}

```

async is another sugar to decorate functions as it always wraps the result of the function into another Promise.

In the case above, async foo() will return a Promise (wrap null, the result of foo()), so that we can

```
async bar() {
    await foo();
    await promise1;
    await promise2;
    await promise3;
}
```
Those promises will be executed in order.

Back to the question.

```
async executeContract(contractId, argument, functionArgument) {
    // ...

    const promise = new Promise((resolve, reject) => {
      this.ledgerClient.executeContract(
          request,
          this.metadata,
          (err, response) => {
            if (err) {
              reject(err);
            } else {
              const jsonResponse = response.toObject();
              jsonResponse.result = JSON.parse(jsonResponse.result);
              resolve(jsonResponse);
            }
          },
      );
    });

    return this._executePromise(promise);
}

async _executePromise(promise) {
    try {
      return await promise;
    } catch (e) {
      const status = this._parseStatusFromError(e);
      if (status) {
        throw new ClientError(status.code, status.message);
      } else {
        throw new ClientError(
            StatusCode.UNKNOWN_TRANSACTION_STATUS,
            e.message,
        );
      }
    }
}
```

In _executePromise, first, we await (trigger to execute) the given promise (passed by executeContract) which wraps underneath protobuf functions,
then, if any error occurs, we pack the error into a ClientError and throw it.
If everything goes well, then _executePromise returns the Promise which wraps the result of protobuf function in runtime.
It means, this._executePromise(promise) will return a Promise to executeContract,
and executeContract will return this Promise to the upper layer, then the upper layer can choose to use await to execute it in order.

```
async upperlayer() {
    try {
        const response = await executeContract();
    } catch (e) {
        // ...
    } 
}
```

or use the traditional, asynchronized way

```
upperlayer() {
    executeContract().then(response => {
        // handle response
    }).catch(e => {
        // ...
    });
}
```

Reply:
When we use async/await pattern, then the caller, user can choose to use asynchronized way or synchronized way to make the code flow in order.
If we use the `callback` way (which originally provided by protobuf),
then the user either uses asynchronized way (the callback hell) to make them in order, or the user wraps them into Promise by themself.

That's why I want to implement with async/await pattern.
The user can easily choose the way to call those interfaces
```
try {
    const response = await clientService.registerCertificate()
    const response2 = await clientService.executeContract()
} catch (error) {
    ...
}
```
or
```
clientService.registerCertificate().then(response => {
    clientService.executeContract().then(response2 => {
        ...
    }).catch(error2 => {
        ...
    });
}.catch(error => { ... });
```

Reply:
I cannot say it is the best practice.
But, as a user, I prefer to use an async function than to use a function which needs me to pass a callback function because I might encounter callback hell.

Reply:
@scalarindetail Thank you very much, I get better understanding to it, but I'm not fully sure.

So my concrete question is,
this code
```
async _executePromise(promise) {
      return await promise;
}
```
and this code
```
 _executePromise(promise) {
      return promise;
}
```
are producing different results ?
Since they both return promise, it seems like it still gives users the same choice.

Reply:
> @scalarindetail Thank you very much, I get better understanding to it, but I'm not fully sure.
> 
> So my concrete question is,
> this code
> 
> ```
> async _executePromise(promise) {
>       return await promise;
> }
> ```
> 
> and this code
> 
> ```
>  _executePromise(promise) {
>       return promise;
> }
> ```
> 
> are producing different results ?
> Since they both return promise, it seems like it still gives users the same choice.

Ah, I got your question.

Yes.
These two function return promise.
If the case above they are identical.

However, in our case, it is

```
  async _executePromise(promise) {
    try {
      return await promise;
    } catch (e) {
      const status = this._parseStatusFromError(e);
      if (status) {
        throw new ClientError(status.code, status.message);
      } else {
        throw new ClientError(
            StatusCode.UNKNOWN_TRANSACTION_STATUS,
            e.message,
        );
      }
    }
  }
```
In runtime, before return promise, we trigger promise first and handle any occurred error.
Without async, the promise will always be returned without executed first.
And, because `return promise` will not throw any errors, so the flow will never go into the `catch`.

Reply:
@scalarindetail Ah, I got it. Thank you !

Reply:
Thanks !